x/tx/decode/decode.go

```diff
@@ -137,48 +137,50 @@ func (d *Decoder) Decode(txBytes []byte) (*DecodedTx, error) {
 }
 
 // Hash implements the interface for the Tx interface.
-func (dtx *DecodedTx) Hash() [32]byte {
+func (dtx DecodedTx) Hash() [32]byte {
 	if !dtx.cachedHashed {
-		dtx.computeHashAndBytes()
+		dtx = dtx.computeHashAndBytes()
 	}
 	return dtx.cachedHash
 }
 
-func (dtx *DecodedTx) GetGasLimit() (uint64, error) {
-	if dtx == nil || dtx.Tx == nil || dtx.Tx.AuthInfo == nil || dtx.Tx.AuthInfo.Fee == nil {
+func (dtx DecodedTx) GetGasLimit() (uint64, error) {
+	if dtx.Tx == nil || dtx.Tx.AuthInfo == nil || dtx.Tx.AuthInfo.Fee == nil {
 		return 0, errors.New("gas limit not available or one or more required fields are nil")
 	}
 	return dtx.Tx.AuthInfo.Fee.GasLimit, nil
 }
 
-func (dtx *DecodedTx) GetMessages() ([]proto.Message, error) {
-	if dtx == nil || dtx.Messages == nil {
+func (dtx DecodedTx) GetMessages() ([]proto.Message, error) {
+	if len(dtx.Messages) == 0 || dtx.Messages == nil {
```

Comment:
we can remove the dtx.Messages == nil check, since it's already covered by dtx.Messages == 0.